azure-kusto-data/azure/kusto/data/_kusto_client.py

```diff
@@ -285,50 +270,40 @@ def to_dataframe(self):
+    _tables_kinds = {
         "QueryResult": WellKnownDataSet.PrimaryResult,
         "@ExtendedProperties": WellKnownDataSet.QueryProperties,
         "QueryStatus": WellKnownDataSet.QueryCompletionInformation,
     }
 
     def __init__(self, json_response):
-        if len(json_response["Tables"]) <= 2:
-            json_response["Tables"][0]["TableKind"] = WellKnownDataSet.PrimaryResult
-            json_response["Tables"][0]["TableId"] = 0
-
-            if len(json_response["Tables"]) == 2:
-                json_response["Tables"][1]["TableKind"] = WellKnownDataSet.QueryProperties
-                json_response["Tables"][1]["TableId"] = 1
-            self.tables = [_KustoResultTable(t) for t in json_response["Tables"]]
-            self.tables_count = len(json_response["Tables"])
+        super(_KustoResponseDataSetV1, self).__init__(json_response["Tables"])
+        self._status_column = "StatusDescription"
+        self._crid_column = "ClientActivityId"
+        self._error_column = "Severity"
+        if self.tables_count <= 2:
+            self.tables[0].table_kind = WellKnownDataSet.PrimaryResult
+            self.tables[0].table_id = 0
+
+            if self.tables_count == 2:
+                self.tables[1].table_kind = WellKnownDataSet.QueryProperties
+                self.tables[1].table_id = 1
         else:
-            json_tables = json_response["Tables"][:-1]
-            toc = _KustoResultTable(json_response["Tables"][-1])
-            for i in range(0, len(json_tables)):
-                json_table = json_tables[i]
-                json_table["TableName"] = toc[i]["Name"]
-                json_table["TableId"] = toc[i]["Id"]
-                json_table["TableKind"] = self._v1_tables_kinds[toc[i]["Kind"]]
-                self.tables.append(_KustoResultTable(json_table))
-
-        super(_KustoResponseDataSetV1, self).__init__()
+            toc = self.tables[-1]
+            for i in range(self.tables_count - 1):
+                self.tables[i].table_name = toc[i]["Name"]
+                self.tables[i].table_id = toc[i]["Id"]
+                self.tables[i].table_kind = self._tables_kinds[toc[i]["Kind"]]
 
 
 class _KustoResponseDataSetV2(_KustoResponseDataSet):
-
-    _status_column = "Payload"
-    _error_column = "Level"
-    _crid_column = "ClientRequestId"
-
     def __init__(self, json_response):
-        json_response = [t for t in json_response if t["FrameType"] == "DataTable"]
-        for json_table in json_response:
-            json_table["TableKind"] = WellKnownDataSet[json_table["TableKind"]]
-        self.tables = [_KustoResultTable(json_table) for json_table in json_response]
-        super(_KustoResponseDataSetV2, self).__init__()
+        super(_KustoResponseDataSetV2, self).__init__(
+            [t for t in json_response if t["FrameType"] == "DataTable"]
+        )
+        self._status_column = "Payload"
+        self._error_column = "Level"
+        self._crid_column = "ClientRequestId"
```

Comment:
These assignments may not be necessary - the class level variables you had were fine. (Even during `__init__`, as base classes are initialized they will see the most-derived class variables when they look at `self.<name>`. It's only if they do `_KustoResponseDataSetV1.<name>` that they'll get the exact value on that type.)